### PR TITLE
[WIP] Adding basic Multitask Models

### PR DESCRIPTION
**Pull request by rbharath**
This PR ports over multitask models for pytorch over from DeepChem contrib and adds some basic tests.

### CODE REVIEW

**Comment by VIGS25** [comment]
One cool thing about PyTorch is gradient accumulation - so you can feed the model with smaller batch sizes, but accumulate the gradients for a few steps, allowing an effectively higher batch size

File: torchchem/models/multitask_classification.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+    assert n_layers > 0, 'Must have some layers defined.'
+
+    prev_layer_size = self.n_features
+    self.W_list = []
+    self.b_list = []
+    for i in range(n_layers):
+      W_init = np.random.normal(0, weight_init_stddevs[i],
+                                (prev_layer_size, layer_sizes[i]))
+      #W_init = torch.cuda.FloatTensor(W_init)
+      W_init = torch.FloatTensor(W_init)
```
Comment (VIGS25) [inline torchchem/models/multitask_classification.py]: https://pytorch.org/docs/stable/nn.html#torch.nn.ParameterList

`nn.Parameter` and `nn.ParameterList` are your friends here.

Comment (rbharath) [reply torchchem/models/multitask_classification.py]: As a quick question here, what's best practice in PyTorch models for writing a model that can run on both CPU and GPU? Is there an easy way to do a device type check and run either `W_init = torch.cuda.FlooatTensor` or `W_init = torch.FloatTesnor` depending on GPU or CPU?

Comment (VIGS25) [reply torchchem/models/multitask_classification.py]: The scripts I run usually look like this:
```
DEVICE = "cuda" if torch.cuda.is_available() else "cpu"

model = ModelName(**config, device=DEVICE)
model.to(DEVICE)
```

`model.to(DEVICE)` will change all my submodules (`nn.Module`), and `nn.Parameter (nn.ParameterList, nn.ParameterDict)` to cpu/gpu depending on what I have. 

I also use a method on my Dataset or Model class, that allows me to move the input tensors to CPU or GPU.

Comment (VIGS25) [reply torchchem/models/multitask_classification.py]: Otherwise, I am not aware of any standard practice. I used this for my thesis and I find it works correctly without any modifications.

**Comment by rbharath** [comment]
@VIGS25 Thanks for the pointer! Do you have an example of how to do gradient accumulation correctly?

**Comment by VIGS25** [comment]
```
def _train_step(self, inputs: Tuple[Tuple[torch.Tensor, ...], ...],
                    step_count: int, **kwargs) -> Dict[str, float]:
        """Runs a train step.

        Parameters
        ----------
        inputs: tuple of tuples of torch.Tensors
            Inputs to the WLNDisconnect forward pass
        optimizer: torch.optim.Optimizer:
            optimizer used for gradient computation
        """
        total_loss, metrics = self.model.train_step(*inputs)
        assert torch.isfinite(total_loss).all()

        accum_every = kwargs.get('accum_every', None)
        if accum_every is not None:
            apply_grad = (step_count % accum_every) == 0
            total_loss /= accum_every
            total_loss.backward()

            if apply_grad:
                if "clip_norm" in kwargs:
                    nn.utils.clip_grad_norm_(self.model.parameters(), kwargs["clip_norm"])

                self.optimizer.step()
                self.optimizer.zero_grad()

                if step_count % self.print_every == 0:
                    if torch.cuda.is_available():
                        alloc_memory = torch.cuda.memory_allocated() / 1024.0 / 1024.0
                        cached_memory = torch.cuda.memory_cached() / 1024.0 / 1024.0
                        print(f"Memory: Allocated: {alloc_memory:.3f} MB, Cache: {cached_memory:.3f} MB")
                        sys.stdout.flush()

                if torch.cuda.is_available():
                    torch.cuda.empty_cache()
```

This is how I do it for my thesis right now. You only call `optimizer.step()` after the required number of steps have passed. But you keep using `loss.backward()`, because that keeps adding the gradients.

File: torchchem/models/multitask_classification.py (hunk 1 above)
Comment (VIGS25) [reply torchchem/models/multitask_classification.py]: This is the member function I use for converting input tensors to device
```
def to_device(self, tensors: Union[List, torch.Tensor]) -> Union[List, torch.Tensor]:
        """Converts all inputs to the device used.
        Parameters
        ----------
        tensors: Union[List, torch.Tensor],
            Tensors to convert to model device. The tensors can be either a
            single tensor or an iterable of tensors.
        """
        if isinstance(tensors, list) or isinstance(tensors, tuple):
            tensors = [tensor.to(self.device, non_blocking=True) for tensor in tensors]
            return tensors
        elif isinstance(tensors, torch.Tensor):
            return tensors.to(self.device, non_blocking=True)
        else:
            raise ValueError(f"Tensors of type {type(tensors)} unsupported")
```
This can be a method in our `BaseModel` class and all other DeepChem models can inherit from it.

**Comment by VIGS25** [comment]
If you want to create new tensors which are not parameters in the model, then PyTorch has functions for that - https://pytorch.org/docs/stable/tensors.html#torch.Tensor
This allows you to retain the device of existing tensors that you create the new tensor(s) from

**Comment by rbharath** [comment]
I'm going to merge in this open PR so we get things more setup in the main repo. Can continue the work started here in subsequent PRs
